filebeat/beat/filebeat.go

```diff
@@ -101,17 +112,18 @@ func (fb *Filebeat) Cleanup(b *beat.Beat) error {
 // Stop is called on exit for cleanup
 func (fb *Filebeat) Stop() {
 
-	// Stop harvesters
-	// Stop prospectors
+	logp.Info("Stopping filebeat")
+	// Stop crawler -> stop prospectors -> stop harvesters
+	fb.cralwer.Stop()
```

Comment:
stopping must ripple through the layers. The `done` queue can be either shared globally or we're using done queue per layer.

In harvester there are effectively 2 places the harvester can block:
1. reading from file
2. forwarding event to spooler

Part 2 is quite simple by using select when publishing. Part 1 should be possible by closing the file from another goroutine. You can get check actual stack-trace if filebeat shutdown is hanging by running filebeat with `-httpprof :6060` and pointing your browser to http://localhost:6060/debug/pprof